src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// To quiet analyzers asking to change code from dotnet/runtime style to dotnet/roslyn style
+// <auto-generated/>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed partial class SolutionState
+{
+    /// <summary>
+    /// String comparer for file paths that caches the last result of the comparison to avoid expensive rehashing of the
+    /// same string over and over again.
+    /// </summary>
+    private sealed class CachingFilePathComparer : IEqualityComparer<string>
+    {
+        /// <summary>
+        /// Note: this insensitive comparer is busted on many systems.  But we do things this way for compat with the logic
+        /// we've had on windows since forever.
+        /// </summary>
+        private static readonly StringComparer s_comparer = StringComparer.OrdinalIgnoreCase;
+
+        public static readonly CachingFilePathComparer Instance = new();
+
+        /// <summary>
+        /// Lock to protect the the last string and hash code we computed.  `enableThreadOwnerTracking: false` as we
+        /// don't need that tracking, and it substantially speeds up the spin lock (removing 0.7% cpu from solution load
+        /// scenario).
+        /// </summary>
+        private SpinLock _lock = new(enableThreadOwnerTracking: false);
+        private string? _lastString;
+        private int _lastHashCode;
+
+        private CachingFilePathComparer()
+        {
+        }
+
+        public bool Equals(string? x, string? y)
+            => s_comparer.Equals(x, y);
+
+        public int GetHashCode([DisallowNull] string obj)
+        {
+            string? lastString;
+            int lastHashCode;
+
+            var lockTaken = false;
+            try
+            {
+                _lock.Enter(ref lockTaken);
+                lastString = _lastString;
+                lastHashCode = _lastHashCode;
+            }
+            finally
+            {
+                if (lockTaken)
+                    _lock.Exit();
+            }
+
+            if (ReferenceEquals(lastString, obj))
```

Comment:
>if (ReferenceEquals(lastString, obj)) [](http://example.com/codeflow?start=12&length=37)

Would it make things ugly to check _lastString against obj and only taking the lock if ReferenceEqual?